consensus/system_contract/system_contract.go

```diff
@@ -0,0 +1,94 @@
+package system_contract
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/params"
+	"github.com/scroll-tech/go-ethereum/rollup/sync_service"
+)
+
+const (
+	defaultSyncInterval = 10 * time.Second
+)
+
+// SystemContract
+type SystemContract struct {
+	config *params.SystemContractConfig // Consensus engine configuration parameters
+	client sync_service.EthClient
+
+	signerAddressL1 common.Address // Address of the signer stored in L1 System Contract
+
+	signer common.Address // Ethereum address of the signing key
+	signFn SignerFn       // Signer function to authorize hashes with
+	lock   sync.RWMutex   // Protects the signer and proposals fields
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// New creates a SystemContract consensus engine with the initial
+// signers set to the ones provided by the user.
+func New(ctx context.Context, config *params.SystemContractConfig, client sync_service.EthClient) *SystemContract {
+	ctx, cancel := context.WithCancel(ctx)
+	address, err := client.StorageAt(ctx, config.SystemContractAddress, config.SystemContractSlot, nil)
+	if err != nil {
+		log.Error("failed to get signer address from L1 System Contract", "err", err)
+	}
+	systemContract := &SystemContract{
+		config:          config,
+		client:          client,
+		signerAddressL1: common.BytesToAddress(address),
+
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	systemContract.Start()
+	return systemContract
+}
+
+// Authorize injects a private key into the consensus engine to mint new blocks
+// with.
+func (s *SystemContract) Authorize(signer common.Address, signFn SignerFn) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.signer = signer
+	s.signFn = signFn
+}
+
+func (s *SystemContract) Start() {
+	log.Info("starting SystemContract")
+	go func() {
+		syncTicker := time.NewTicker(defaultSyncInterval)
+		defer syncTicker.Stop()
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			default:
+			}
+			select {
+			case <-s.ctx.Done():
+				return
+			case <-syncTicker.C:
+				address, err := s.client.StorageAt(s.ctx, s.config.SystemContractAddress, s.config.SystemContractSlot, nil)
+				if err != nil {
+					log.Error("failed to get signer address from L1 System Contract", "err", err)
+				}
+				s.lock.Lock()
```

Comment:
This is much better than before but I still think we might be locking a critical path unnecessarily as in probably 99.9% of the queries the value will not change. Therefore, and since this is the only thread that would ever change the value, I think it would be better to do an `RLock` first, read and compare the value and only if it changed, do a `Lock`. This is a bit more verbose and ugly but imo this is justified here